File: console/src/test/java/com/jraska/console/ConsoleTest.java
package com.jraska.console;

import android.content.Context;
import android.graphics.Color;
import android.text.SpannableString;
import android.text.style.ForegroundColorSpan;
import android.view.View;

import org.junit.After;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.Mockito;
import org.robolectric.RobolectricTestRunner;
import org.robolectric.RuntimeEnvironment;
import org.robolectric.annotation.Config;

import java.lang.ref.WeakReference;

import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.Matchers.isA;
import static org.mockito.Mockito.verify;

@RunWith(RobolectricTestRunner.class)
@Config(constants = BuildConfig.class, sdk = 26)
public class ConsoleTest {
  //region Fields

  private Console console;

  //endregion

  //region Setup methods

  @Before
  public void setUp() {
    console = new Console(getAppContext());
  }

  @After
  public void tearDown() {
    Console.buffer.setSize(ConsoleBuffer.MAX_BUFFER_SIZE);
    Console.clear();
    Console.consoles.clear();
  }

  //endregion

  //region Test Methods

  @Test
  public void testWrite()  {
    String testText = "asd5a6das7";
    Console.write(testText);

    assertThat(console.getConsoleText()).contains(testText);
  }

  @Test
  public void testWriteLine()  {
    String testText = "657ad52jh";
    Console.writeLine(testText);

    assertThat(console.getConsoleText()).contains(testText);
    assertThat(console.getConsoleText()).endsWith(Console.END_LINE);
  }

  @Test
  public void testWriteSpannable()  {
    SpannableString spannableString = new SpannableString("123456789");
    spannableString.setSpan(new ForegroundColorSpan(Color.RED), 5, 8, 0);

    Console.write(spannableString);

    assertThat(console.getConsoleText()).isEqualTo(spannableString.toString());
  }

  @Test
  public void testWriteLineSpannable()  {
    SpannableString spannableString = new SpannableString("123456789");
    spannableString.setSpan(new ForegroundColorSpan(Color.BLUE), 0, 4, 0);

    Console.writeLine(spannableString);

    assertThat(console.getConsoleText()).contains(spannableString);
    assertThat(console.getConsoleText()).endsWith(Console.END_LINE);
  }

  @Test
  public void testClear()  {
    Console.write(new SpannableString("123456789"));
    Console.writeLine("as6ad77asd8");

    Console.clear();

    assertThat(console.getConsoleText()).isEqualTo("");
  }

  @Test
  public void whenTextLongerThenBufferSize_printedTextIsShortened()  {
    Console.buffer.setSize(5);

    Console.write("123456789");

    assertThat(console.getConsoleText()).isEqualTo("56789");

    for (int i = 0; i < 5; i++) {
      Console.writeLine("");
    }

    assertThat(console.getConsoleText()).isEqualTo("\n\n\n\n\n");
  }

  @Test
  public void whenBufferSizeChanges_textIsShortened()  {
    Console.write("123456789");

    Console.buffer.setSize(6);
    Console.scheduleBufferPrint();

    assertThat(console.getConsoleText()).isEqualTo("456789");
  }

  @Test
  public void whenBufferSizeChanges_thenSpannableShortened()  {
    SpannableString spannableString = new SpannableString("123456789");
    spannableString.setSpan(new ForegroundColorSpan(Color.BLUE), 3, 6, 0);

    Console.write(spannableString);

    Console.buffer.setSize(5);
    Console.scheduleBufferPrint();

    assertThat(console.getConsoleText()).isEqualTo("56789");
  }

  @Test
  public void whenWrittenMultipleTimes_thenScrollDownScheduledOnlyOnce()  {
    Console consoleSpy = Mockito.spy(console);
    Console.consoles.set(0, new WeakReference<>(consoleSpy));
    consoleSpy.measure(0, 0); // simulate next frame

    Console.write("someText");
    Console.write("line");

    verify(consoleSpy).post(isA(Runnable.class));
  }

  @Test
  public void whenNewConsoleViewCreated_thenBufferIsPrinted()  {
    Console.write("text");

    Console newConsole = new Console(getAppContext());
    assertThat(newConsole.getConsoleText()).isEqualTo("text");
  }

  //region View methods failures tests

  @Test(expected = UnsupportedOperationException.class)
  public void testFailsOn_addView()  {
    console.addView(new View(getAppContext()));
  }

  @Test(expected = UnsupportedOperationException.class)
  public void testFailsOn_removeViewAt()  {
    console.removeViewAt(0);
  }

  @Test(expected = UnsupportedOperationException.class)
  public void testFailsOn_removeView()  {
    console.removeView(console.getChildAt(0));
  }

  @Test(expected = UnsupportedOperationException.class)
  public void testFailsOn_removeAllViews()  {
    console.removeAllViews();
  }

  @Test(expected = UnsupportedOperationException.class)
  public void testFailsOn_removeAllViewsInLayout()  {
    console.removeAllViewsInLayout();
  }

  //endregion

  //endregion

  //region Properties

  private Context getAppContext() {
    return RuntimeEnvironment.application;
  }

  //endregion
}

File: console/src/test/java/com/jraska/console/ConsoleTest.kt
package com.jraska.console

import android.graphics.Color
import android.text.SpannableString
import android.text.style.ForegroundColorSpan
import org.assertj.core.api.Assertions.assertThat
import org.junit.After
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.mockito.ArgumentMatchers.isA
import org.mockito.Mockito
import org.mockito.Mockito.verify
import org.robolectric.RobolectricTestRunner
import org.robolectric.RuntimeEnvironment.application
import org.robolectric.annotation.Config
import java.lang.ref.WeakReference
import java.util.concurrent.CountDownLatch
import java.util.concurrent.TimeUnit

@RunWith(RobolectricTestRunner::class)
@Config(sdk = [26])
class ConsoleTest {

  private lateinit var console: Console

  @Before
  fun setUp() {
    console = Console(application)
  }

  @After
  fun tearDown() {
    buffer().setSize(ConsoleBuffer.MAX_BUFFER_SIZE)
    Console.clear()
    Console.controller.consoles.clear()
  }

  @Test
  fun testWrite() {
    val testText = "asd5a6das7"
    Console.write(testText)

    assertThat(console.text).contains(testText)
  }

  @Test
  fun testWriteLine() {
    val testText = "657ad52jh"
    Console.writeLine(testText)

    assertThat(console.text).contains(testText)
    assertThat(console.text).endsWith(ConsoleController.END_LINE)
  }

  @Test
  fun testWriteSpannable() {
    val spannableString = SpannableString("123456789")
    spannableString.setSpan(ForegroundColorSpan(Color.RED), 5, 8, 0)

    Console.write(spannableString)

    assertThat(console.text).isEqualTo(spannableString.toString())
  }

  @Test
  fun testWriteLineSpannable() {
    val spannableString = SpannableString("123456789")
    spannableString.setSpan(ForegroundColorSpan(Color.BLUE), 0, 4, 0)

    Console.writeLine(spannableString)

    assertThat(console.text).contains(spannableString)
    assertThat(console.text).endsWith(ConsoleController.END_LINE)
  }

  @Test
  fun testClear() {
    Console.write(SpannableString("123456789"))
    Console.writeLine("as6ad77asd8")

    Console.clear()

    assertThat(console.text).isEqualTo("")
  }

  @Test
  fun whenTextLongerThenBufferSize_printedTextIsShortened() {
    buffer().setSize(5)

    Console.write("123456789")

    assertThat(console.text).isEqualTo("56789")

    for (i in 0..4) {
      Console.writeLine("")
    }

    assertThat(console.text).isEqualTo("\n\n\n\n\n")
  }

  @Test
  fun whenBufferSizeChanges_textIsShortened() {
    Console.write("123456789")

    buffer().setSize(6)
    Console.controller.scheduleBufferPrint()

    assertThat(console.text).isEqualTo("456789")
  }

  @Test
  fun whenBufferSizeChanges_thenSpannableShortened() {
    val spannableString = SpannableString("123456789")
    spannableString.setSpan(ForegroundColorSpan(Color.BLUE), 3, 6, 0)

    Console.write(spannableString)

    buffer().setSize(5)
    Console.controller.scheduleBufferPrint()

    assertThat(console.text).isEqualTo("56789")
  }

  @Test
  fun whenWrittenMultipleTimes_thenScrollDownScheduledOnlyOnce() {
    val consoleSpy = Mockito.spy(console)
    Console.controller.consoles[0] = WeakReference(consoleSpy)
    consoleSpy.measure(0, 0) // simulate next frame

    Console.write("someText")
    Console.write("line")

    verify(consoleSpy).post(isA(Runnable::class.java))
  }

  @Test
  fun whenNewConsoleViewCreated_thenBufferIsPrinted() {
    Console.write("text")

    val newConsole = Console(application)
    assertThat(newConsole.text).isEqualTo("text")
    assertThat(Console.consoleCount()).isEqualTo(2)
  }

  @Test
  fun survivesProperlyConcurrencyIssues() {
    buffer().setSize(10)

    val finishLatch = CountDownLatch(1)
    val appendRunnable = Runnable {
      for (i in 0..1000) {
        Console.write("textToAppend")
      }

      finishLatch.countDown()
    }
    Thread(appendRunnable).start()

    for (i in 0..1000) {
      Console.writeLine("daa")
    }

    val await = finishLatch.await(1, TimeUnit.SECONDS)
    assertThat(await).isTrue()
  }

  private fun buffer() = Console.controller.buffer
}
